MessageSigning/MessageSigning.md

```diff
@@ -0,0 +1,437 @@
+
+You can find more complete definitions of `protected` and `unprotected` in [RFC 8152 section-3](https://tools.ietf.org/html/rfc8152#section-3) and you can see some the reserved entries (called `Generic_Headers`) in the maps in [RFC 8152 section-3.1](https://tools.ietf.org/html/rfc8152#section-3.1).
+
+Note: `payload` can be `nil`. This means that the payload is known by both the signer and the verifier and therefore doesn't need to be encoded.
+
+For your convenience, the structure is provided here:
+
+```
+empty_or_serialized_map = bstr .cbor header_map / bstr .size 0
+
+header_map = {
+    Generic_Headers,   ; reserved headers (see COSE section 3.1)
+    * label => values  ; any number of int/string labels for application-specific purpose.
+}
+
+Headers = (
+    protected : empty_or_serialized_map,
+    unprotected : header_map
+)
+
+; signature layer
+COSE_Signature =  [
+    Headers,
+    signature : bstr
+]
+
+; if signing with just ONE key
+COSE_Sign1 = [
+    Headers,
+    payload : bstr / nil,
+    signature : bstr
+]
+
+# if signing with >1 key
+COSE_Sign = [
+    Headers,
+    payload : bstr / nil,
+    signatures : [+ COSE_Signature]
+]
+
+signed_message = COSE_SIGN / COSE_Sign1
+```
+
+#### Signing and Verification target format
+
+Instead of signing the full structure, we instead sign the following type which is derived from the structure
+
+```
+Sig_structure = [
+  context : "Signature" / "Signature1" / "CounterSignature",    ; explained below
+  body_protected : empty_or_serialized_map,                     ; protected from layer 1
+  ? sign_protected : empty_or_serialized_map,                   ; not present in Sign1 case
+  external_aad : bstr,                                          ; explanation below
+  payload : bstr
+]
+```
+
+The `external_aad` allows an application to ask the user to sign some extra data but NOT put it inside the COSE structure (only as part of the data to sign). Defaults to `h''`. You can read more about this at [RFC 8152 section-4.3](https://tools.ietf.org/html/rfc8152#section-4.3).
+
+The `context` is meant to encode what structure was used for the COSE request. `CounterSignature` is explained in a later section of this specification.
+
+#### Signing and Verification process
+
+To be able to effectively verify we need two things:
+
+1) P1 - (optional) knowledge of the relation of a public key and a Cardano address
+1) P2 - Knowledge of which algorithm was used to sign
+
+For `P1`, just knowing the public key is not sufficient to know the address (which we need to know if we want to prove ownership of a UTXO or transaction). To resolve this, first we (optionally) add to the header an extra field.
```

Comment:
Why can't the sender put the address in the signed message if they need to convey it?

Reply:
In fact, I'm a little confused by the rationale here. During transaction validation, all that is required to prove ownership of a UTXO is (a signature from) the corresponding public key. Why do we need more here?

Reply:
@SebastienGllmt when it comes purely to the signing / verification procedure, it is purely Ed25519 even for legacy addresses. The chaincode and attributes are needed in order to re-compute the address root so that one can check that the signature was produced by the right private key. 

I am not sure to understand the full rationale here then. If we're looking for at message signing from a more general perspective, then the public key itself is sufficient to verify a signature made from a "legacy" private key.